parachain-upgrade/src/lib.rs

```diff
@@ -178,9 +166,21 @@ impl<T: Trait> ProvideInherent for Module<T> {
 	const INHERENT_IDENTIFIER: InherentIdentifier = INHERENT_IDENTIFIER;
 
 	fn create_inherent(data: &InherentData) -> Option<Self::Call> {
-		let vfp = data.get_data(&INHERENT_IDENTIFIER)
-			.and_then(|r| r.ok_or_else(|| "Validation Function Params inherent data not found".into()))
-			.expect("Gets and decodes vfp inherent data");
+		let vfp: ValidationFunctionParams = data.get_data(&INHERENT_IDENTIFIER)
+			.expect("Gets and decodes vfp inherent data")
+			.expect("Validation Function Params inherent data not found");
```

Comment:
These types of assertions are not allowed in any of our codebases. The approach with panickers is "Prove or remove" - the message of an `expect` statement should be a proof that the expect will not occur. This lets us know which assumptions have been violated when the program encounteres a panic.

There are rare situations in the runtime where we intentionally want to panic during transaction processing or `on_finalize` because we want the block to become invalid. However, this is not one of those situations.

If the inherent is mandatory, I'd recommend returning `None` here but panicking in the `on_finalize` implementation if a `DidUpdate: bool` in storage is not `true`. See what is done in the `frame-timestamp` or `polkadot-runtime-common/parachains` modules - this `DidUpdate` value should never hit disk, so it is an efficient check.

Reply:
Good spot, thanks! db26357 does as you say; there was already an `on_finalize` assertion.